docs: architecture guide (#1335)
finos/git-proxy: docs/Architecture.md

# GitProxy Architecture

This guide explains GitProxy's various components, and how they communicate with each other when performing a `git push`.

As mentioned in [the README](/README.md), GitProxy is an application that intercepts pushes and applies rules/policies to ensure they're compliant. Although a number of policies are available by default, these can be extended by using plugins.

## Overview

GitProxy has several main components:

- Proxy (`/src/proxy`): The actual proxy for Git. Git operations performed by users are intercepted here to apply the relevant **chain**. Also loads **plugins** and adds them to the chain. Runs by default on port `8000`.
  - Chain: A set of **processors** that are applied to an action (i.e. a `git push` operation) before requesting review from a user with permission to approve pushes
  - Processor: AKA `Step`. A specific step in the chain where certain rules are applied. See the [list of default processors](#processors) below for more details.`

>   - Plugin: A custom processor that can be added externally to extend GitProxy's default policies. See the plugin guide for more details.

Reviewer: ```suggestion
  - Plugin: A custom processor that can be added externally to extend GitProxy's default policies. See the [plugin guide](https://git-proxy.finos.org/docs/development/plugins) for more details.
```

Note that `writePack` sets Git's `receive.unpackLimit` to `0`, which forces Git to always create pack files instead of unpacking objects individually.

Source: [/src/proxy/processors/push-action/writePack.ts](/src/proxy/processors/push-action/writePack.ts)

#### `checkHiddenCommits`

Detects "hidden" commits in a push, which is possible if the pack file in the push was tampered in some way.

It calls `git verify-pack` on each of the new `.idx` files found in [`writePack`](#writepack). If any unreferenced commits are present, the push is blocked.

Source: [/src/proxy/processors/push-action/checkHiddenCommits.ts](/src/proxy/processors/push-action/checkHiddenCommits.ts)

#### `checkIfWaitingAuth`

Checks if the action has been authorised (approved by a reviewer). If so, allows the push to continue to the remote. It simply continues chain execution if the push hasn't been approved.

Source: [/src/proxy/processors/push-action/checkIfWaitingAuth.ts](/src/proxy/processors/push-action/checkIfWaitingAuth.ts)

#### `preReceive`

> Allows executing pre-receive hooks from `.sh` scripts located in the `./hooks` directory. **Also allows bypassing the manual approval process.** This enables admins to reuse GitHub enterprise commit policies and provide a seamless experience for contributors who no longer need to wait for approval or be aware of GitProxy intercepting their pushes.

Reviewer: ```suggestion
Allows executing pre-receive hooks from `.sh` scripts located in the `./hooks` directory. **Also allows automating the approval process.** This enables admins to reuse GitHub enterprise commit policies and provide a seamless experience for contributors who no longer need to wait for manual approval or be aware of GitProxy intercepting their pushes.
```

Pre-receive hooks are a feature that allows blocking unwanted commits based on rules described in `.sh` scripts. GitHub provides a set of [sample rules](https://github.com/github/platform-samples/blob/master/pre-receive-hooks) to get started.

This processor will block the push depending on the exit status of the pre-receive hook:

> - Exit status `0`: Sets the push to `autoApproved`, skipping the requirement for subsequent approval. Note that this doesn't affect the other processors, which may still block the push.

Reviewer: ```suggestion
- Exit status `0`: Sets the push to `autoApproved`. If no other processors block the push, the contributor can immediately push again to the upstream repository without waiting for manual approval.
```

Author: Do we let the end user know if their push is ready to push again in this case? I thought the push would go through the processors and then automatically get pushed to the remote in one go. 🤔 

Just want to make sure that there's a clear message displayed so that users don't "wait" for approval when there's no need!

Reviewer 2: @jescalada The original design was the user had to 'repush''  *if* the push was blocked for any manually steps. The reason for this - was didn't want to store any sensitive credential data in GP itself. Once approved the user receives an email or other notification that they can re-push. 

If the initial push is fully auto-approved all the way through (i.e. all checks were automated), then the push should just carry on through and the users receives the usual git response that their push was good - i.e. no notification that ''you can push'' would be required (as, the user is 'done') 

If someone can verify that is the current implementation, as I get my head back into the project.

-Paul

> <!-- Todo: confirm whether pushes that passed the prereceive can still be blocked -->

Reviewer: ```suggestion

```
